### ISSUE TITLE
`smt run` with modified params fails for non numeric types

### ISSUE THREAD

**Issue by ashander** [open-research/sumatra#221]
Specifically, bools do not work (using JSON parameter file) and are instead read in as strings. That is, R sees "false". 

In fact, looking at `commands.py` line 60 this may be a more general issue (i.e., occurring in python as well as R) as there is no attempt to cast command line arguments to boolean types

**Comment by maxalbert** [comment]
Indeed, I've had this problem with Python simulations for a while. See [this discussion](https://groups.google.com/forum/#!topic/sumatra-users/VTHJMRGbsMk) on the mailing list. I thought we had an issue for this already, but this doesn't appear to be the case. Thanks for reporting it!

It would be possible to apply a kludgy ad-hoc fix quite easily, but as Andrew points in the aforementioned thread there are more subtle issues to make this work for all input formats. It's been on my list for quite a while to fix this. I'll see if I can give this a shot over the weekend. (But if you're really keen to give it a try yourself, feel free. :) )

**Comment by ashander** [comment]
Ah, thanks for the pointer to the mailing list. I had looked searched
Issues but didn't realize the mailing list existed!

I agree with your comments there. If the ParameterSet classes deal with
parsing comandline params, things should *just work*. For example, command
line parameters work fine for R executable w/ JSON if one modifies
parse_command_line_parameter to the following:

```python
def json_bool_to_python_bool(value):
    '''converts json strings true or false to bool'''
    if value == 'true':
        return True
    elif value == 'false':
        return False
    else:
        raise ValueError

def parse_command_line_parameter(p):
    pos = p.find('=')
    if pos == -1:
        raise Exception("Not a valid command line parameter. String must be
of form 'name=value'")
    name = p[:pos]
    value = p[pos + 1:]
    if list_pattern.match(value) or tuple_pattern.match(value):
        value = eval(value)
    else:
        for cast in int, float, json_bool_to_python_bool:
            try:
                value = cast(value)
                break
            except ValueError:
                pass
    return {name: value}
```

**Comment by maxalbert** [comment]
True, although this still feels somewhat ad-hoc imho. (Btw, you could use `json.loads` in place of your custom function `json_bool_to_python_bool` to avoid having to do write similar functions for non-boolean types). Also, the key issue is that it shouldn't just work for JSON input, but for all supported input formats. Thus I think it would be worth restructuring the command line parsing code and outsourcing part of the functionality to the various `ParameterSet` classes (because these are designed to know how to deal with input in a particular format).

**Comment by maxalbert** [comment]
I think I've got the beginnings of a PR lying around somewhere. I'll try to dig it up and tidy it up over the weekend (unless you or someone else beat me to it :) ).

**Comment by ashander** [comment]
+1 the code above is super adhoc. Putting the logic in ParameterSet methods and the format library methods like `json.load`, `yaml.load` etc is the way to go. 

One note: seems from a quick look this will require changing the logic in parse_arguments. I'll look out for your PR and use the above hack on my local branch for now ;)

**Comment by ashander** [comment]
One more thing, if you do try the PR this weekend, include this change to `.travis.yml` as maybe hte travis build infrastructure has changed: https://github.com/ashander/sumatra/commit/40029f1caa0d6ea2ac87ea39e1fee97f606f64f4

**Comment by ashander** [comment]
After this discussion, I went ahead and opened a PR making the change for JSON and moving the parsing to the parameterset later.

If your PR ends up being incompatible, or more thoughtful, I won't be offended if you scrap this one.

**Comment by ashander** [comment]
This is fixed for JSON in #223 

Should still fail for other formats -- eg YAML, Config -- but see that PR as template to add handling for other formats.
